ksmt-core/src/main/kotlin/io/ksmt/expr/rewrite/simplify/KBvExprSimplifier.kt

```diff
@@ -114,1001 +91,899 @@ interface KBvExprSimplifier : KExprSimplifierBase {
+            simplifier = { v, rotation -> postRewriteBvRotateLeftExpr(v, rotation) }
+        )
 
-            // (x >>> x) ==> 0
-            if (arg == shift) {
-                return@simplifyExpr bvZero(size).uncheckedCast()
-            }
+    fun <T : KBvSort> KContext.preprocess(expr: KBvRotateRightIndexedExpr<T>): KExpr<T> = expr
 
-            mkBvLogicalShiftRightExprNoSimplify(arg, shift)
-        }
+    // (rotateRight a x) ==> (rotateLeft a (- size x))
+    fun <T : KBvSort> KContext.postRewriteBvRotateRightIndexedExpr(rotation: Int, value: KExpr<T>): KExpr<T> =
+        simplifyBvRotateRightIndexedExpr(rotation, value)
 
-    override fun <T : KBvSort> transform(expr: KBvArithShiftRightExpr<T>): KExpr<T> =
-        simplifyExpr(expr, expr.arg, expr.shift) { arg, shift ->
-            simplifyBvArithShiftRightExpr(arg, shift)
-        }
+    override fun <T : KBvSort> transform(expr: KBvRotateRightIndexedExpr<T>): KExpr<T> =
+        simplifyExpr(
+            expr = expr,
+            a0 = expr.value,
+            preprocess = { preprocess(it) },
+            simplifier = { postRewriteBvRotateRightIndexedExpr(expr.rotationNumber, it) }
+        )
 
-    // (repeat a x) ==> (concat a a ..[x].. a)
-    override fun transform(expr: KBvRepeatExpr): KExpr<KBvSort> = simplifyExpr(expr, expr.value) { arg ->
-        val repeats = arrayListOf<KExpr<KBvSort>>()
-        repeat(expr.repeatNumber) {
-            repeats += arg
-        }
+    fun <T : KBvSort> KContext.preprocess(expr: KBvRotateRightExpr<T>): KExpr<T> = expr
 
-        if (repeats.size == 0) {
-            return@simplifyExpr mkBvRepeatExprNoSimplify(expr.repeatNumber, arg)
-        }
+    fun <T : KBvSort> KContext.postRewriteBvRotateRightExpr(value: KExpr<T>, rotation: KExpr<T>): KExpr<T> =
+        simplifyBvRotateRightExpr(value, rotation)
 
-        return@simplifyExpr rewrite(SimplifierFlatBvConcatExpr(ctx, expr.sort, repeats))
-    }
+    override fun <T : KBvSort> transform(expr: KBvRotateRightExpr<T>): KExpr<T> =
+        simplifyExpr(
+            expr = expr,
+            a0 = expr.arg,
+            a1 = expr.rotation,
+            preprocess = { preprocess(it) },
+            simplifier = { v, rotation -> postRewriteBvRotateRightExpr(v, rotation) }
+        )
 
-    // (zeroext a) ==> (concat 0 a)
-    override fun transform(expr: KBvZeroExtensionExpr): KExpr<KBvSort> = simplifyExpr(expr, expr.value) { arg ->
-        if (expr.extensionSize == 0) {
-            return@simplifyExpr arg
-        }
+    fun <T : KBvSort> KContext.preprocess(expr: KBvAddNoOverflowExpr<T>): KExpr<KBoolSort> = expr
 
-        if (arg is KBitVecValue<*>) {
-            return@simplifyExpr arg.zeroExtension(expr.extensionSize.toUInt()).uncheckedCast()
-        }
+    fun <T : KBvSort> KContext.postRewriteBvAddNoOverflowExpr(
+        lhs: KExpr<T>,
+        rhs: KExpr<T>,
+        isSigned: Boolean
+    ): KExpr<KBoolSort> = rewriteBvAddNoOverflowExpr(lhs, rhs, isSigned)
```

Comment:
Here we have to use special rewriting rules (`rewriteBvAddExpr`, `rewriteAnd`, etc. ).